lib/rgit/cli.rb

```diff
@@ -55,13 +258,275 @@ def format_options(option_hash)
+      # dir != parent check is needed since user may "re-init" a repo
+      if (git_parent) && directory != git_parent
+        init_subrepo(git_parent, subrepo_parent, directory)
+      else
+        init_repo(directory, options)
+      end
     end
 
     desc "test", "A test function"
     def test(*args)
-      puts args
+      j = args.class
+    end
+  end
+end
+
+module Impl
+  VERSION = Rgit::VERSION
+  include Wrappers
+  require "pty"
+  def format_options(option_hash)
+    result = ""
+    option_hash.each { |k, v|
+      key_str = " --#{k.to_s.gsub("_", "-")}"
+      case v # v is truthy in all cases except: nil, false
+        when [true, false].include?(v) # it's a boolean
+          result << key_str
+        when String
+          result << key_str << "=#{v}"
+        when Fixnum
+          result << key_str << "=#{v}"
+        when Hash
+          v.each { |key, val|
+            result << key_str << " #{key}=#{val}"
+          }
+      end
+    }
+    result
+  end
+
+  def lowest_repo_above(start_dir)
+    res_dir = nil
+    start_dir.realpath.ascend { |dir|
+      if dir.has_child? ".git"
+        res_dir = dir
+        break
+      end
+    }
+    res_dir
+  end
+
+  def lowest_gitrepo_above(start_dir)
+    res_dir = nil
+    start_dir.realpath.ascend{ |dir|
+      if dir.has_child? ".git"
+        res_dir = dir
+        break
+      end
+    }
+    res_dir
+  end
+
+  def lowest_above(start_dir)
+    lowest_gitrepo = lowest_gitrepo_above(start_dir)
+    lowest_subrepo = nil
+    if lowest_gitrepo
+      branches = Branches.new(lowest_gitrepo)
+      lowest_subrepo = branches.path_for_branch(branches.branch_for_path(start_dir))
+    end
+    [lowest_gitrepo, lowest_subrepo]
+  end
+
+  # Calling git & shells
+  def git_command(cmd, opt_str, dir)
+    run_in_shell("git #{cmd} #{opt_str}", dir)
+  end
+
+  def run_in_shell(cmd, dir)
+    output = `cd #{dir}; #{cmd}; cd; cd - ;`
+    output
+  end
+
+  # Opens a virtual shell at the specified
+  # dir and runs the given cmd
+  def run_in_pty(cmd, dir)
+    result = ''
+    PTY.spawn("cd #{dir.realpath}; #{cmd}") do |stdout, stdin, pid|
+      begin
+        stdout.each { |line| result += line }
+      rescue Errno::EIO #Done getting output
+        result
+      end
+    end
+    result
+  end
+
+  # Transactions
+  def stash_unstash(dir)
+    git_command("stash", "", dir)
+    yield
+    git_command("stash", "pop --index", dir)
+  end
+
+  # git branch inside is the top branch of the group we were just in
+  def uncap_cap!(git_dir)
+    # grab info we need to restore state, and remove the cap
+    branches = Branches.new(git_dir)
+    group_head = branches.current_group
+    current_branch = branches.current_branch
+    git_command("reset", "--hard HEAD~1", git_dir)
+    yield branches # give block the branches object so it knows where it is
+    branches.update
+    opt_str = ""
+    branches[group_head].each{|branch| opt_str << branch << " "}
```

Comment:
Space missing to the left of {.<br>Space between { and | missing.<br>Space missing inside }.